src/AdsImporter/Importer.php

```diff
@@ -0,0 +1,245 @@
+                ++$this->importerResult->blocks;
+            } catch (CommandException $ex) {
+                if ($ex->getCode() !== CommandError::GET_BLOCK_INFO_UNAVAILABLE) {
+                    $this->addExceptionToLog($ex, 'get_block', ['block' => $blockId]);
+                }
+            }
+
+
+            $startTime += $this->blockSeqTime;
+            $blockId = NumericalTransformation::decToHex($startTime);
+        } while ($startTime <= $endTime);
+
+        return $this->importerResult;
+    }
+
+    private function getStartTime(): int
+    {
+        $from = $this->databaseMigration->getNewestBlockTime();
+
+        if ($from) {
+            return $from + $this->blockSeqTime;
+        }
+
+        return $this->genesisTime;
+    }
+
+    private function updateNodes(): void
+    {
+        try {
+            $blockResponse = $this->client->getBlock();
+        } catch (CommandException $ex) {
+            throw new AdsClientException('Cannot proceed importing data');
+        }
+
+        $nodes = $blockResponse->getBlock()->getNodes();
+
+        /** @var Node $node */
+        foreach ($nodes as $node) {
+            if ($node->getId() === '0000') { // special node
+                continue;
+            }
+
+            $this->updateAccounts($node);
+
+            $this->databaseMigration->addOrUpdateNode($node);
+            ++$this->importerResult->nodes;
+        }
+    }
+
+    /**
+     * @param Node $node
+     *
+     * todo how we can generate hash (or something) of a node to not update anything if there is no changes
+     */
+    private function updateAccounts(Node $node): void
+    {
+        $accountResponse = $this->client->getAccounts($node->getId());
+        $accounts = $accountResponse->getAccounts(); // todo check if `getAccounts` method always returns array
+
+        /** @var Account $account */
+        foreach ($accounts as $account) {
+            $this->databaseMigration->addOrUpdateAccount($account, $node);
+            ++$this->importerResult->accounts;
+        }
+    }
+
+    private function addPackagesFromBlock(Block $block): int
+    {
+        $blockTransactionsCount = 0;
+
+        try {
+            $packagesResponse = $this->client->getPackageList($block->getId());
+            $packages = $packagesResponse->getPackages();
+
+            /** @var Package $package */
+            foreach ($packages as $package) {
+                $package->generateId();
+
+                $transactionsCount = $this->addTransactionsFromPackage($package, $block);
+
+                $package->setTransactionCount($transactionsCount);
+                $this->databaseMigration->addPackage($package, $block);
+
+                ++$this->importerResult->packages;
+                $blockTransactionsCount += $transactionsCount;
+            }
+        } catch (CommandException $ex) {
+            // @todo if there is no packages for a block ads client should NOT throw an exception
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1135)